### PR TITLE
Nail down requireds

### PR DESCRIPTION
**Pull request by echeng06**
Ended up requiring specifying the type for _all_ gets on the raw data. This should both enforce and make explicit what the expected type of each raw/derived attribute is, and also lets us remove a lot of castings/checking of Nones. 

Expected typing is based on the DED/RDD, although I expect we might snag a few edge cases that require us to loosen things here or there. 

Also fixed tests - I ended up removing the E2E UDS one, since it's so entangled with other variables/forms it's more trouble than its worth to maintain.

### CODE REVIEW

File: nacc_attribute_deriver/attributes/base/uds_namespace.py [hunk 1]
```diff
@@ -9,46 +9,47 @@
...
+        self, table: SymbolTable, required: Optional[frozenset[str]] = None
+    ) -> None:
         """Check that this is a UDS form."""
-        super().__init__(table=table)
+        if required is None:
+            required = frozenset()
 
-        module = self.get_value("module")
-        if not module or module.upper() != "UDS":
+        default_required = ["module", "packet", "formver", "birthmo", "birthyr"]
```
Comment (bjkeller) [inline nacc_attribute_deriver/attributes/base/uds_namespace.py:19]: is these version specific? are we going to have a different list with v4?

File: nacc_attribute_deriver/attributes/mqt/cognitive.py [hunk 2]
```diff
@@ -21,8 +21,28 @@ class CognitiveAttributeCollection(AttributeCollection):
...
-        self.__derived = DerivedNamespace(table=table)
+        self.__uds = UDSNamespace(
+            table,
+            required=frozenset(
+                [
+                    "cdrglob",
+                ]
+            ),
+        )
+        self.__derived = DerivedNamespace(
```
Comment (bjkeller) [inline nacc_attribute_deriver/attributes/mqt/cognitive.py:32]: Is this collection always used in a context where the derived variables always exist?

if so, we need to split the collection

File: nacc_attribute_deriver/schema/operation.py [hunk 3]
```diff
@@ -45,12 +45,10 @@ def get_list_type(expression_type: type) -> type:
 
 def get_date_tagged_type(expression_type: type) -> type:
     if hasattr(expression_type, "__pydantic_generic_metadata__"):
-        origin = expression_type.__pydantic_generic_metadata__["origin"]  # type: ignore
+        origin = expression_type.__pydantic_generic_metadata__["origin"]
```
Comment (bjkeller) [inline nacc_attribute_deriver/schema/operation.py:48]: We lost the type ignore pragmas in this file. They are all things we can't do anything about, so I'd like to keep them

File: rule_reflection/inspect_rule_types.py [hunk 4]
```diff
@@ -31,8 +31,8 @@ def type_string(type_object: type) -> str:
         return "None"
 
     if hasattr(type_object, "__pydantic_generic_metadata__"):
-        origin = type_object.__pydantic_generic_metadata__["origin"]  # type: ignore
-        args = type_object.__pydantic_generic_metadata__["args"]  # type: ignore
+        origin = type_object.__pydantic_generic_metadata__["origin"]
```
Comment (bjkeller) [inline rule_reflection/inspect_rule_types.py:34]: we lost all of the type ignore pragmas in this file. I'd like to keep them

File: tests/schema/test_attribute_types.py [hunk 5]
```diff
@@ -62,19 +62,19 @@ def test_set(self):
         value = "blah"
         operation.evaluate(table=table, value=value, attribute="single-value")
         assert operation.attribute_type(type(value)) is List[type(value)]
-        assert type(table.get("single-value")[0]) is type(value)  # type: ignore
+        assert type(table.get("single-value")[0]) is type(value)
```
Comment (bjkeller) [inline tests/schema/test_attribute_types.py:65]: please restore the type ignore pragmas in this file

**Review by bjkeller — CHANGES_REQUESTED** [review]
This looks great. There are a couple of minor changes and questions

Though, some of the files you didn't change lost the type ignore pragmas that I had added to deal with cases where we can't do anything about the typing. Please restore those.

File: nacc_attribute_deriver/attributes/base/uds_namespace.py (hunk 1 above)
Comment (echeng06) [reply nacc_attribute_deriver/attributes/base/uds_namespace.py:19]: These specific ones are not version specific, but we could end up with version-specific lists. NP would also have very version-specific lists.

File: nacc_attribute_deriver/attributes/mqt/cognitive.py (hunk 2 above)
Comment (echeng06) [reply nacc_attribute_deriver/attributes/mqt/cognitive.py:32]: In general the assumption is MQT-variables run after all NACC-variables have been curated. We currently handle this implicitly by the rule order in `curation_rules.csv`, but I agree it's not intuitive and we may need to add something more explicit.
